blocksync/params.go

```diff
@@ -0,0 +1,113 @@
+package blocksync
+
+import (
+	"time"
+)
+
+// BlockPoolParams holds dynamically calculated parameters for the block pool
+type BlockPoolParams struct {
+	maxPendingPerPeer int
+	retryTimeout      time.Duration
+	requestersLimit   int
+	blockSizeBuffer   *BlockStats
+	maxRequesters     int
+
+	// Cached calculated values for logging
+	avgBlockSize     float64 // kept for logging purposes
+	maxBlockSize     float64 // used for calculations
+	peerBasedLimit   int
+	memoryBasedLimit int
+	numSamples       int
+}
+
+// NewBlockPoolParams creates a new BlockPoolParams with the given configuration
+func NewBlockPoolParams(blockSizeBuffer *BlockStats, maxRequesters int) *BlockPoolParams {
+	params := &BlockPoolParams{
+		blockSizeBuffer: blockSizeBuffer,
+		maxRequesters:   maxRequesters,
+	}
+	params.recalculate(0)
+	return params
+}
+
+// recalculate updates all parameters based on current conditions
+// numPeers is passed in since it's external state from BlockPool
+func (p *BlockPoolParams) recalculate(numPeers int) {
+	p.avgBlockSize = p.blockSizeBuffer.GetAverage()
+	p.maxBlockSize = p.blockSizeBuffer.GetMax()
+	p.numSamples = p.blockSizeBuffer.Size()
+
+	// Use defaults if not enough samples
+	if p.numSamples < minSamplesForDynamic {
+		p.maxPendingPerPeer = defaultMaxPendingRequestsPerPeer
+		p.retryTimeout = time.Duration(defaultRetrySeconds * float64(time.Second))
+	} else {
+		// Use max block size for calculations (worst-case planning)
+		p.maxPendingPerPeer = p.calculateMaxPending(p.maxBlockSize)
```

Comment:
Why we're using max observed block size? IMHO this is not a worst case. Using max allowed block size is actually the worst case.

Reply:
Using max observed block will be more conservative and allow us to have larger timeouts for the future blocks, otherwise if we get some blocks at 0 and some blocks 128 we will have timeouts that would be to low for large blocks. We can also use median for that. But still, I don't want to complicate this right now